src/dwt.rs

```diff
@@ -0,0 +1,219 @@
+//! Debug and trace and stuff
+
+use crate::rcc::Clocks;
+use crate::time::Hertz;
+use cortex_m::peripheral::{DCB, DWT};
+use embedded_hal::blocking::delay::{DelayMs, DelayUs};
+
+pub trait DwtExt {
+    fn constrain(self, dcb: DCB, clocks: Clocks) -> Dwt;
+}
+impl DwtExt for DWT {
+    /// Enable trace unit and cycle counter
+    fn constrain(mut self, mut dcb: DCB, clocks: Clocks) -> Dwt {
+        cortex_m::interrupt::free(|_| {
+            // Enable fmc clock
+            dcb.enable_trace();
+
+            self.enable_cycle_counter();
+        });
+
+        Dwt {
+            dwt: self,
+            dcb,
+            clocks,
+        }
+    }
+}
+
+pub struct Dwt {
+    dwt: DWT,
+    dcb: DCB,
+    clocks: Clocks,
+}
+impl Dwt {
+    /// Release the dwt and dcb control
+    /// NOTE all instances of Delay and StopWatch become invalid after this
+    pub fn release(self) -> (DWT, DCB) {
```

Comment:
Yeah, a proper implementation could borrow Dwt, bit the api will become annoying. What about putting this function unsafe? Even if that's not really unsafe, should only cause deadlocks.

Reply:
Changed to unsafe